### PR TITLE
feat(examples): add Qwen3-32B Tulu-3 finetune recipes

### PR DESCRIPTION
**Pull request by athitten**
# What does this PR do ?

Adds dense **Qwen3-32B** Tulu-3 SFT example recipes as a context-parallelism **cp1-vs-cp2 parity pair**. Qwen3-32B has no custom NeMo model class, so these load the stock HF `Qwen3ForCausalLM` directly. Each conversation is pinned to a fixed **2048-token** sequence (unpacked) so CP has a real, uniform sequence to shard, and the two arms consume identical data (`shuffle: false`) so their loss curves must overlap.

# Changelog

- **`examples/llm_finetune/qwen/qwen3_32b_tulu3.yaml`** — cp1 baseline (FSDP2, `cp_size=1`).
- **`examples/llm_finetune/qwen/qwen3_32b_tulu3_cp2.yaml`** — cp2 arm (`cp_size=2`, `local_batch_size=2`); identical to cp1 apart from the CP settings, so `global_batch_size=64` and `grad_accum=8` stay fixed.
- **`examples/llm_finetune/qwen/qwen3_tulu3_chat_template.jinja`** — generation-block chat template. The stock Qwen3 template rewrites earlier turns (drops `<think>` from non-final turns), so `ChatDataset` cannot build the answer-only loss mask by prefix diffing (raises since #3024). This template wraps assistant spans in `{% generation %}` so the tokenizer emits the mask directly.
- **Loss** = `FusedLinearCrossEntropy` (with `output_hidden_states: true`) to avoid the `[batch, seq, vocab]` fp32 logit upcast that OOMs `MaskedCrossEntropy` on 32B.
- **Fixed seqlen 2048** — dataset `truncation: true` + `seq_length: 2048` caps the ~0.4% long tail; collater `pad_seq_len_divisible: 2048` pads every batch up to 2048 (padding is masked, so it is transparent to the loss).
- **`distributed.defer_fsdp_grad_sync: false`** — required for 32B full-FT to fit on 8×80GB; the FSDP2 default accumulates the full unsharded 32B gradient (~64 GiB bf16) across grad-accum steps and OOMs.
- Packing stays off: with `cp_size>1` the only usable attention is SDPA (CP intercepts `scaled_dot_product_attention`), and packing under CP over SDPA is unsupported for a stock HF model (rationale in the cp2 recipe comment).

# Validation — cp1 vs cp2 loss parity (100 steps)

Both arms run identical data (`shuffle: false`, `global_batch_size: 64`) on 8×H100 at seqlen 2048:

```bash
# cp1
automodel examples/llm_finetune/qwen/qwen3_32b_tulu3.yaml \
  --nproc-per-node 8 --step_scheduler.max_steps 100 --dataset.split "train[:20000]"
# cp2
automodel examples/llm_finetune/qwen/qwen3_32b_tulu3_cp2.yaml \
  --nproc-per-node 8 --step_scheduler.max_steps 100 --dataset.split "train[:20000]"
```

(`train[:20000]` is a fixed slice so both arms see the same samples; the committed recipe uses the full `split: train`.)

| metric | value |
|---|---|
| steps compared | 100 |
| mean \|Δ\| | 0.00013 |
| **max \|Δ\|** | **0.00080** (at step 2) |
| step 0 | cp1 `1.2508` / cp2 `1.2506` |
| step 50 | cp1 `1.1413` / cp2 `1.1413` |
| step 99 | cp1 `1.0625` / cp2 `1.0625` |

The cp1 and cp2 loss curves are numerically identical (differences are float reduction-order noise) → context parallelism is correct on Qwen3-32B. A 10-step run agreed to `max|Δ| = 0.0005`.

**wandb (cp1 vs cp2 overlay):** 
- cp1 (`qwen3_32b_cp1_tulu3`): [https://wandb.ai/nvidia/qwen3_32b/runs/mrrl2407?nw=nwuserathittenaman](https://wandb.ai/nvidia/qwen3_32b/runs/mrrl2407?nw=nwuserathittenaman)
- cp2 (`qwen3_32b_cp2_tulu3`): [https://wandb.ai/nvidia/qwen3_32b/runs/orev2rb0?nw=nwuserathittenaman](https://wandb.ai/nvidia/qwen3_32b/runs/orev2rb0?nw=nwuserathittenaman)
Screenshot 2026-07-17 at 5 31 14 PM

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests? — N/A: example configs only. They auto-enroll as CI functional recipes and are validated by the cp1-vs-cp2 parity run above.
- [x] Did you add or update any necessary documentation? — recipe header comments + chat-template rationale.

# Additional Information

- Related to # (issue)

### CODE REVIEW

**Review by HuiyingLi — APPROVED** [review]
Thank you so much!
